### ISSUE TITLE
The Unboundid Library still tries to make the connection when the connection is about to age out

### ISSUE THREAD

**Issue by nsyed4** [pingidentity/ldapsdk#77]
Application using the Unboundid LDAP Library: Java based/Play Framework.
Unboundid Library Version: unboundid-ldapsdk: 4.0.12
Unboundid Library Connects to: ForgeRock OpenDJ Version 5.5
To Establish the connection Pool:
`final LDAPConnectionPool ldapConnectionPool =
                new LDAPConnectionPool(serverSet, new EXTERNALBindRequest(""), initialConnectionCount, maximumConnectionCount, new LoggingPostConnectProcessor());`

`final Duration maxConnectionAge = Duration.ofMinutes(4);
ldapConnectionPool.setMaxConnectionAgeMillis(maxConnectionAge.toMillis());`

Many times we are seeing that exactly at 4th minute, the LDAPConnection disconnects. This is expected as per the settings. But it also disconnects while there is ADD operation going on.

Example below is the OpenDJ Server log:
`[19/Jan/2020:02:33:14] ADD conn=2898 op=28 result=118 message="Client Unbind" etime=121
[19/Jan/2020:02:33:14] DISCONNECT conn=2898 reason="Client Unbind"
[19/Jan/2020:02:33:14] UNBIND conn=2898 op=29 msgID=31`

On the Client side:
`2020-01-19 02:33:14,312 level="ERROR" message="LDAPResult(resultCode=82 (local error), messageID=30, diagnosticMessage='The connection to the directory server was closed while waiting for a response to an asynchronous request.')" `

Why is the library closing the connection abruptly instead of waiting for the response? Is there a setting for drain time that new requests are not processed on this connection when connection the connection is going to get closed.

What if we setMaxAge to 0? A value of zero indicates that no maximum age should be enforced. We have an initial pool size of 5 and maximum pool size of 15.

**Comment by dirmgr** [comment]
A connection pool should never close (or do anything else with) a connection that has been checked out of the pool. The pool wouldn't even have a reference to that connection for things like health checking or evaluating the maximum connection age.

If you're seeing a pooled connection get closed while it's being used, then the only reason I could think that might happen would be if the connection has already been checked back into the pool, but you still have a reference to that connection and try to use it. For example, the following could result in the behavior that you're reporting:

```
LDAPConnection connection = pool.getConnection();
pool.releaseConnection(connection);
connection.add(addRequest);
```

Is there a chance that the add request is being issued over a connection that might have already been released back to the pool?

**Comment by dirmgr** [comment]
Actually, from taking a closer look at the error message you're getting back, it seems like you're using asynchronous operations. Those are very difficult to get right when you're using connection pooling, specifically because it's very easy for a connection to be inadvertently used for multiple concurrent things that don't know about each other, and in that case, one thread using a could do something that would adversely impact other connections.

The only way to safely use asynchronous operations with connections from a connection pool is to check a connection out of a pool, perform the asynchronous processing while it's checked out, and then only check that connection back into the pool after all asynchronous processing has completed. If you initiate an asynchronous operation with a connection that is checkout of of a pool, but then check the connection back into the pool before the asynchronous processing is complete, then you can run into exactly the scenario that you're describing.

**Comment by dirmgr** [comment]
And to answer your question about a maximum connection age of zero, that is documented in the javadoc for the setMaxConnectionAgeMillis method: "A value of zero indicates that no maximum age should be enforced." That is the default behavior for the LDAP SDK, and the LDAP SDK will inherently use a value of zero unless you override it.

**Comment by nsyed4** [comment]
Thanks @dirmgr. This was very helpful. Yes, we are using asynchronous operations.

**Comment by nsyed4** [comment]
Hi @dirmgr 

Currently, the client gets connection from the pool and makes a request to LDAP. While the client is waiting for the response from the server; because the max connection age on the pool is set to 4 minutes, the connection is killed at 4th minute, resulting in the result=118 message="Client Unbind" error. 

`setMaxConnectionAgeMillis(long maxConnectionAge)(Specifies the maximum length of time in milliseconds that a connection in this pool may be established before it should be closed and replaced with another connection.)`


Is it possible that the library doesn't get the connection from the pool if it is about to expire OR remove the connections from the pool that are about to expire. If the client doesn't use "about to expire" connection, then there will be no chance that they make the request and the connection expires before the response comes back.

**Comment by dirmgr** [comment]
A connection pool should never attempt to close a connection because of the maximum age while it’s checked out of the pool. The pool only looks at a connection’s age when it’s performing health checking on connections, and it only performs health checking on connections that are checked into the pool. And you should never attempt to use a connection that isn’t checked into the pool.

Are you perhaps keeping a reference to a connection after it’s been checked back into the pool? Are there still asynchronous operations in progress when you check a connection back into the pool? Both of those are definitely things that you should not do. When issuing asynchronous requests on a connection checked out of a pool, you should make sure that all of those operations have completed (or at least, all of them that you care about) before checking the connection back into the pool. Once a connection is checked back into the pool, it’s fair game to be closed by the pool or affected by some other operation that tries to use the connection (for example, attempting a bind on the connection, which would cause all outstanding operations to be canceled).

If you’re sure that the connection is checked out of the pool, then you could try enabling debugging support (see the [com.unboundid.util.Debug](https://docs.ldap.com/ldap-sdk/docs/javadoc/index.html?com/unboundid/util/Debug.html) javadoc for details) for more information about the context in which the connection is being closed. If you’re still using the 4.0.12 release, you might want to consider upgrading, since we made some improvements in connection pool debugging support in the 4.0.13 release. If you can capture debugging information when the connection is closed, we should be able to use it to better understand the cause of the closure.

**Comment by nsyed4** [comment]
Hi,

Here is the code that is run by multiple threads. Also, we never did setSynchronousMode

     LDAPConnectionPool pool = poolProvider.getLdapPool();
     LDAPConnection connection = null;
        try {
            pool = poolProvider.getLdapPool();
            connection = pool.getConnection();
            final AsyncRequestID asyncRequestId = connection.asyncSearch(request);
            pool.releaseConnection(connection);
        } catch (final LDAPException ex) {
           TryEffect.f((LDAPException lex) -> promise.failure(lex)).f(ex);
            pool.releaseConnectionAfterException(connection, ex);
        }
/*Omitting the Listener code that processes the response...*/


Do you think that the above code snippet could be a problem because pool.releaseConnection is executed before the response comes back. Now, pool.releaseConnection checks the maxAge (which we set to 4 min), and if it is greater than 4 minutes, connection.terminate is executed. If the connection is terminated before the response comes back (that is, request is still processing in Open DJ), that's when DJ throws an error like: 
`[19/Jan/2020:02:33:14] ADD conn=2898 op=28 result=118 message="Client Unbind" etime=121 [19/Jan/2020:02:33:14] DISCONNECT conn=2898 reason="Client Unbind" `

**Comment by dirmgr** [comment]
Yes, the problem is that you’re calling releaseConnection before the search has completed. Releasing the connection tells the pool that it’s free to do whatever it wants with that connection, including closing it or handing for processing with other operations.

If you’re using asynchronous operations, you must ensure that any outstanding operations you care about have completed before releasing the connection back to the pool.
